core/domain/rights_manager.py

```diff
@@ -962,7 +1097,7 @@ def _deassign_role(committer, removed_user_id, activity_id, activity_type):
             'This user does not have any role in %s with ID %s'
             % (activity_type, activity_id))
 
-    assignee_username = user_services.get_usernames(removed_user_id)[0]
+    assignee_username = user_services.get_usernames([removed_user_id])[0]
```

Comment:
Weird, how did this work before, or pass tests? (Is it possible that we now have data corruption on the server)?

/cc @kevintab95

Reply:
Can you update us about this @kevintab95 !!

Reply:
@sahiljoster32 Just to confirm, this is a question for you, though @kevintab95 may have some insight. (That's what "cc" means -- otherwise I would have addressed it to him directly.)

Reply:
I think Previously when we are passing any `removed_user_id`, `user_services.get_usernames()` always returns None ( because it can give Optional values too ) and that none is assigned to `assignee_username`. 

So, if we see line `1126`, there we are handling the case of None and assigning `assignee_username` with `ANONYMOUS`. I think that's the reason why backend tests are passing. Because while saving `activity_rights` ( in line 1140 ), the generated comit_message only contains `ANONYMOUS` values.

Reply:
Confirmed! checked all the test cases and in all test cases previously all `assignee_username`s are None.

Reply:
Hm, I think we need to add more test cases then -- and you might want to talk to @kevintab95 directly about this since there may be a serious bug here that has affected the datastore.

Reply:
found more info!! 
if we see the `_assign_role` method, there we are doing the same thing but we are using `user_services.get_username`

https://github.com/oppia/oppia/blob/5c6c538ec98b9de3f3d95c1222d66bd2a31bda2a/core/domain/rights_manager.py#L819
and internally `user_services.get_username` is implemented same as `user_services.get_usernames([removed_user_id])[0]` 
https://github.com/oppia/oppia/blob/5c6c538ec98b9de3f3d95c1222d66bd2a31bda2a/core/domain/user_services.py#L1248
I'll write an email to @kevintab95 about this but in mean time can you please take a pass on other comments ?

Reply:
From email thread:

I looked into this and it seems like the "deassign role" method works fine e.g. it does successfully remove roles from the exp rights model. However, as Sahil pointed out, the usernames get recorded wrongly in the ExplorationRightsSnapshotMetadataModel entities in the backend -- we store the username in the commit message. So instead of saying something like "Remove abc from role voice artist for exploration", it records "Remove ANONYMOUS from role ...". Although we don't surface this anywhere, it does imply that the data on our backend might not be representing the right thing. We can fix this by writing a job that edits the commit message in the model by replacing ANONYMOUS with the relevant username corresponding to the "removed_user_id" field in "commit_cmds".


Created https://github.com/oppia/oppia/issues/15779 to fix the commit messages.

Reply:
@sahiljoster32 Just to follow up on this thread, we should have a backend test that prevents the previous error from happening. Did you write one, and if so, could you please point me to it?

Thanks!

Reply:
To de_assign the role we have to first assign the role using the method `assign_role_for_exploration`. So, if the user id does not exist at the time of assignment then an error is displayed. 
Also, if we just pass any `user_id` to de_assign, then the error is displayed just before we hit the line `1125`. Because in this case, every condition results in false because `user_id` is not assigned before so line 1122 throws an error.

But to cover the backend coverage I have written a test with hypothetical case where i mock the function and return None . Name of the test case `test_deassign_editor_is_successful_with_commit_message_having_anonymous`

Reply:
Previously, users are assigned correctly but "Function is implemented wrong", Now function implementation is corrected.

Reply:
Sorry, I'm a bit confused by this. I don't think the issue is that the user ID does not exist at the time of assignment. I think the issue is that assignee_username used to be calculated wrongly, and that means that the resulting commit message is wrong.

So all your test needs to do is to verify that the commit message that is sent to _save_activity_rights is correct now. You can do this by, for example, retrieving the relevant model with the commit message after this function is called, and examining it.

Also @sahiljoster32 I have a question. Why didn't `user_services.get_usernames(removed_user_id)` previously detect that the argument was a string rather than a list(str), as it should have been instead?

Reply:
> So all your test needs to do is to verify that the commit message that is sent to _save_activity_rights is correct now.

Test is added as `test_deassign_editor_is_successful_with_all_valid_commit_messages` in line `868`. In this test, I have checked both the assigning and the de-assigning commit messages.

> Why didn't user_services.get_usernames(removed_user_id) previously detect that the argument was a string rather than a list(str)?

Let me explain it with a simple example first:
```python
def print_elements(x: Union[str, List[str]) -> None:
    for item in x:
        print(item)

# calling `print_elements` with list ~ call 1
print_elements(['hi', 'bie'])
# output of call 1
hi
bie

# calling `print_elements` with string ~ call 2
print_elements('hi')
# output of call 2
h
i
```

Now, somewhat similar is happend with `user_services.get_usernames` previously:
```python
def get_usernames(user_ids) -> List[Optional[str]]:
        for index, user_id in enumerate(user_ids):
             # sudo code just for explaining.
             if user_id exists return user_name:            # Keep this line's user_id in mind. ----------x
                 print(username)
             else: 
                 None

user_id_a = 'abcdefg'     # This user_id has sahil username.

# calling `get_usernames` with list ~ call 1
get_usernames(['user_id_a'])[0]
# output of call 1
sahil                                                       # Because the above user_id is 'abcdefg' and it has username sahil.


# calling `get_usernames` with string ~ call 2
get_usernames('user_id_a')[0]
# output of call 2                                   # Because above user_id has 's', 'a', 'h', 'i' and 'l' values and there is no user_id for these values.
None
```

I hope it make things clearer :)

Reply:
For the last question, I get that (re the behaviour of str vs list). I'm just asking why this wasn't caught by mypy typing (should it have been?)

Reply:
Actually, this error is caught by MYPY but previously MYPY is not throwing error because this file is not excluded from MYPY_DENY_LIST. So, Mypy is not checking for this file. As soon as we removed rights_manager.py from deny_list MyPY checked argument types from both sides and due to the difference in types Mypy throws an `incompatible argument type error`, from that point we get know that something is wrong.

Does that answer your question?

Reply:
Yup. Thanks!